feat(api): express stalls in a recoverable way

sfoster1 opened the pull request:
Commands that use `move_to_well`, `move_to_coordinates`,  `move_to_addressable_area`, and `move_relative` now will return stalls as DefinedErrors, which means they can be hooked into error recovery en masse.

This just leaves move labware.

Closes EXEC-831

## Reviews
- Feel like it paid off?

## Testing
- Do some stalls and make sure they end up with defined errors

## Further work and questions
- I think we should probably handle errors during `aspirate`'s automatic prepare for aspirate at the beginning and probably the same with drop tip but we didn't handle it at all so far so I don't know

sfoster1 committed:
feat(api): express stalls in a recoverable way

Commands that use move_to_well now will return stalls as DefinedErrors,
which means they can be hooked into error recovery en masse.

Closes EXEC-831

sfoster1 committed:
feat(api): handle stalls in move_relative

TamarZanzouri reviewed (approved):
That was quick! Thank you!

TamarZanzouri commented:
one small note, I worry that the error construction will now be scattered in a  bunch of places, but it makes sense to add this logic in one place so I think its ok.

SyntaxColoring commented on api/tests/opentrons/protocol_engine/commands/test_dispense.py:
```diff
@@ -234,3 +238,60 @@ async def test_overpressure_error(
+    model_utils: ModelUtils,
+    state_view: StateView,
+) -> None:
+    """It should return an overpressure error if the hardware API indicates that."""
```
```suggestion
    """It should return a stall error if the hardware API indicates that."""
```

sfoster1 committed:
feat(api): handle stalls in addressable area moves

Both move to addressable are and move to addressable area for drop tip
handle stalls now.

SyntaxColoring commented on api/src/opentrons/protocol_engine/commands/movement_common.py:
```diff
@@ -79,6 +92,22 @@ class MovementMixin(BaseModel):
     )
 
 
+class StallOrCollisionError(ErrorOccurrence):
+    """Returned when the machine detects that axis encoders are reading a different position than expected.
+
+    All axes are stopped at the point where the error was encountered.
+
+    The next thing to move the machine must account for the robot not having a valid estimate
+    of its position. It should be a `home` or `unsafe/updatePositionEstimators`.
+    """
+
+    isDefined: bool = True
+    errorType: Literal["stallOrCollision"] = "stallOrCollision"
+
+    errorCode: str = ErrorCodes.STALL_OR_COLLISION_DETECTED.value.code
+    detail: str = ErrorCodes.STALL_OR_COLLISION_DETECTED.value.detail
```
Thanks for the docstring.

SyntaxColoring reviewed (approved):
Nice.

Definitely paid off—way better than a global `try`/`catch`. Thank you!

sfoster1 committed:
feat(api): handle stalls in move_to_coordinates

sfoster1 committed:
Update api/tests/opentrons/protocol_engine/commands/test_dispense.py

Co-authored-by: Max Marrone <max@opentrons.com>

sfoster1 commented:
> one small note, I worry that the error construction will now be scattered in a bunch of places, but it makes sense to add this logic in one place so I think its ok.

Well, that's the goal with the micro-operations - now they're not scattered, they're all in one place for each kind of micro-operation
